html/man/man3/MediaLibrary.3
.TH "MediaLibrary" 3 "Sat May 16 2015" "Blokbox" \" -*- nroff -*-
.ad l
.nh
.SH NAME
MediaLibrary \- 
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <medialibrary\&.h>\fP
.PP
Inherits QWidget\&.
.SS "Signals"

.in +1c
.ti -1c
.RI "void \fBaddMediaToPlayList\fP (QString)"
.br
.in -1c
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBMediaLibrary\fP (QWidget *parent=0)"
.br
.ti -1c
.RI "\fB~MediaLibrary\fP ()"
.br
.ti -1c
.RI "void \fBaddToSearchPath\fP (QString path)"
.br
.ti -1c
.RI "QSet< QString > \fBgetSearchPath\fP (void)"
.br
.RI "\fIgetSearchPath returns a QSet containing QStrings with song paths \fP"
.in -1c
.SS "Protected Member Functions"

.in +1c
.ti -1c
.RI "void \fBresizeEvent\fP (QResizeEvent *e)"
.br
.ti -1c
.RI "void \fBsetTableTag\fP (QString property, QString value)"
.br
.in -1c
.SS "Private Slots"

.in +1c
.ti -1c
.RI "void \fBonSelectItem\fP (QModelIndex index)"
.br
.ti -1c
.RI "void \fBonDoubleClickItem\fP (QModelIndex index)"
.br
.in -1c
.SS "Private Member Functions"

.in +1c
.ti -1c
.RI "QStandardItem * \fBsearchArtist\fP (QString artist)"
.br
.ti -1c
.RI "QStandardItem * \fBsearchAlbum\fP (QStandardItem *artistItem, QString album)"
.br
.ti -1c
.RI "QStandardItem * \fBsearchSong\fP (QStandardItem *albumItem, QString song)"
.br
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "Ui::MediaLibrary * \fBui\fP"
.br
.ti -1c
.RI "QMediaPlayer * \fBauxplayer\fP"
.br
.ti -1c
.RI "QSet< QString > \fBsearchPath\fP"
.br
.RI "\fIsearchPath stores the search path for seeking songs \fP"
.ti -1c
.RI "QStandardItemModel * \fBtreeModel\fP"
.br
.ti -1c
.RI "QStandardItemModel * \fBtableModel\fP"
.br
.in -1c
.SH "Constructor & Destructor Documentation"
.PP 
.SS "MediaLibrary::MediaLibrary (QWidget * parent = \fC0\fP)\fC [explicit]\fP"

.SS "MediaLibrary::~MediaLibrary ()"

.SH "Member Function Documentation"
.PP 
.SS "void MediaLibrary::addMediaToPlayList (QString)\fC [signal]\fP"

.SS "void MediaLibrary::addToSearchPath (QString path)"

.SS "QSet<QString> MediaLibrary::getSearchPath (void)"

.PP
getSearchPath returns a QSet containing QStrings with song paths 
.PP
\fBReturns:\fP
.RS 4
the set of paths 
.RE
.PP

.SS "void MediaLibrary::onDoubleClickItem (QModelIndex index)\fC [private]\fP, \fC [slot]\fP"

.SS "void MediaLibrary::onSelectItem (QModelIndex index)\fC [private]\fP, \fC [slot]\fP"

.SS "void MediaLibrary::resizeEvent (QResizeEvent * e)\fC [protected]\fP"

.SS "QStandardItem * MediaLibrary::searchAlbum (QStandardItem * artistItem, QString album)\fC [private]\fP"

.SS "QStandardItem * MediaLibrary::searchArtist (QString artist)\fC [private]\fP"

.SS "QStandardItem * MediaLibrary::searchSong (QStandardItem * albumItem, QString song)\fC [private]\fP"

.SS "void MediaLibrary::setTableTag (QString property, QString value)\fC [protected]\fP"

.SH "Member Data Documentation"
.PP 
.SS "QMediaPlayer* MediaLibrary::auxplayer\fC [private]\fP"

.SS "QSet<QString> MediaLibrary::searchPath\fC [private]\fP"

.PP
searchPath stores the search path for seeking songs 
.SS "QStandardItemModel * MediaLibrary::tableModel\fC [private]\fP"

.SS "QStandardItemModel* MediaLibrary::treeModel\fC [private]\fP"

.SS "Ui::MediaLibrary* MediaLibrary::ui\fC [private]\fP"


.SH "Author"
.PP 
Generated automatically by Doxygen for Blokbox from the source code\&.


medialibrary.cpp
#include "medialibrary.h"
#include "ui_medialibrary.h"

MediaLibrary::MediaLibrary(QWidget *parent) :
  QWidget(parent),
  ui(new Ui::MediaLibrary)
{
  ui->setupUi(this);
  treeModel = new QStandardItemModel;
  ui->treeViewLibrary->setModel(treeModel);
  ui->treeViewLibrary->setEditTriggers(QAbstractItemView::NoEditTriggers);
  connect(ui->treeViewLibrary,SIGNAL(clicked(QModelIndex)),
          SLOT(onSelectItem(QModelIndex)));
  connect(ui->treeViewLibrary,SIGNAL(doubleClicked(QModelIndex)),
          SLOT(onDoubleClickItem(QModelIndex)));
  QStringList header;
  header << "Data" << "Path";
  treeModel->setHorizontalHeaderLabels(header);
  ui->treeViewLibrary->sortByColumn(0);
  ui->treeViewLibrary->setSortingEnabled(true);

  // prepares the tableView for viewing tags
  tableModel= new QStandardItemModel(1,2);
  tableModel->setHorizontalHeaderItem(0,new QStandardItem(QString("Propriedade")));
  tableModel->setHorizontalHeaderItem(1,new QStandardItem(QString("Valor")));
  ui->tableViewPlaylist ->verticalHeader()->setVisible(false);
  ui->tableViewPlaylist->horizontalHeader()->setSectionResizeMode(QHeaderView::Interactive);
  ui->tableViewPlaylist->horizontalHeader()->setStretchLastSection(true);
  ui->tableViewPlaylist->setSelectionBehavior(QAbstractItemView::SelectRows);
  ui->tableViewPlaylist->setAlternatingRowColors(true);
  ui->tableViewPlaylist->setEditTriggers(QAbstractItemView::NoEditTriggers);
  ui->tableViewPlaylist->setModel(tableModel);
  ui->tableViewPlaylist->setFont(QFont("Times"));

}

void MediaLibrary::resizeEvent(QResizeEvent *e){
  Q_UNUSED(e);
  ui->treeViewLibrary->setColumnWidth(0,width()/2);
}

MediaLibrary::~MediaLibrary()
{
  delete ui;
}

void MediaLibrary::onDoubleClickItem(QModelIndex index){
  if(!index.model()->hasChildren(index)){
    emit(addMediaToPlayList(index.parent().child(index.row(),1).data().toString()));
  }
}

void MediaLibrary::onSelectItem(QModelIndex index){
  QString path, artist, album, title;
  if(index.model()->hasChildren(index))
    return;
  path = index.parent().child(index.row(),1).data().toString();
#ifdef USE_TAGLIB
  TagLib::FileRef ref(path.toStdString().c_str());
  if(!ref.isNull() && ref.tag()){
    tag=ref.tag();
    if(tag!=NULL){
      artist = tag->artist().toCString();
      if(artist == "")
        artist = "Desconhecido";
      album = tag->album().toCString();
      if(album == "")
        album = "Desconhecido";
      title = tag->title().toCString();
      if(title == "")
        title = "Desconhecido";
      setTableTag(QString("Artista"), artist);
      setTableTag("Album", album);
      setTableTag("Titulo",title);
    }
  }
#endif
}

void MediaLibrary::setTableTag(QString property, QString value){
  QList <QStandardItem*> list;
  QStandardItem *item;
  list = tableModel->findItems(property);
  if(!list.empty()){
    int row;
    row = list.at(0)->row();
    tableModel->item(row,0)->setText(property);
    tableModel->item(row,1)->setText(value);
  }
  else{
    int count;
    count = tableModel->rowCount();
    count++;
    item = new QStandardItem(property);
    tableModel->setItem(count,0,item);
    item = new QStandardItem(value);
    tableModel->setItem(count,1,item);
    tableModel->sortRole();
    tableModel->sort(0);
  }
}

QStandardItem *MediaLibrary::searchArtist(QString artist){
  QList<QStandardItem*> found;
  found = treeModel->findItems(artist,0);
  if(found.size() == 0)
    return NULL;
  else
    return found.at(0);
/*
  for(int i=0; i<model->rowCount(); i++){
    if(model->item(i,0)->text() == artist)
      return model->item(i);
  }
  return NULL;*/
}

QStandardItem* MediaLibrary::searchAlbum(QStandardItem *artistItem, QString album){
  for(int i=0; i<artistItem->rowCount(); i++){
    if(artistItem->child(i)->text() == album)
      return artistItem->child(i);
  }
  return NULL;
}

QStandardItem* MediaLibrary::searchSong(QStandardItem *albumItem, QString song){
  for(int i=0; i<albumItem->rowCount(); i++){
    if(albumItem->child(i)->text() == song)
      return albumItem->child(i);
  }
  return NULL;
}

void MediaLibrary::addToSearchPath(QString path){
  QDir dir;
  // check if the path is already on the set
  if(searchPath.contains(path)){
    // if yes, return
    return;
  }
  // insert path on searchPath
  searchPath.insert(path);
  dir.setPath(path);
  QStack<QString> stack;
  QString fullPath, artist, album, title;
  QStandardItem *artistItem, *albumItem, *songItem, *urlItem, *emptyItem;
  QList<QStandardItem*> list;
  stack.push(dir.absolutePath());
  while (!stack.isEmpty()) {
    QString sSubdir = stack.pop();
    QDir subdir(sSubdir);

#ifdef USE_TAGLIB
    // Check for the files.
    QStringList entries = subdir.entryList(QStringList() << "*mp3" << "*flac", QDir::Files);
    for (int i = 0; i < entries.size(); i++) {
      fullPath = (subdir.path()+"/"+entries[i]);
      TagLib::FileRef ref(fullPath.toStdString().c_str());
      if(!ref.isNull() && ref.tag()){
        tag=ref.tag();
        if(tag!=NULL){
          artist = tag->artist().toCString();
          if(artist == "")
            artist = "Desconhecido";
          album = tag->album().toCString();
          if(album == "")
            album = "Desconhecido";
          title = tag->title().toCString();
          if(title == "")
            title = "Desconhecido";

          // find the artist
          artistItem = searchArtist(artist);
          // if it does not exist, create it
          if(artistItem==NULL){
            artistItem = new QStandardItem(artist);
            emptyItem = new QStandardItem;
            list.clear();
            list << artistItem << emptyItem;
            treeModel->appendRow(list);
          }

          // search for the album
          albumItem = searchAlbum(artistItem, album);
          // if it does not exist, create it
          if(albumItem==NULL){
            albumItem = new QStandardItem(album);
            artistItem->appendRow(albumItem);
          }

          // search for the song
          songItem = searchSong(albumItem, title);
          // if it does not exist, create it
          if(songItem==NULL){
            list.clear();
            songItem = new QStandardItem(title);
//            albumItem->appendRow(songItem);
            urlItem = new QStandardItem(fullPath);
            list << songItem << urlItem;
            albumItem->appendRow(list);
          }
        }
      }
    }
#endif
    QFileInfoList infoEntries = subdir.entryInfoList(QStringList(),
                                                     QDir::AllDirs | QDir::NoSymLinks | QDir::NoDotAndDotDot);
    for (int i = 0; i < infoEntries.size(); i++) {
      QFileInfo& item = infoEntries[i];
      stack.push(item.absoluteFilePath());
    }
  }
  ui->treeViewLibrary->sortByColumn(0);
}
